### PR TITLE
Maaike

### PR DESCRIPTION
**Pull request by maaike-slt**
yes

### CODE REVIEW

File: libft/inc/libft.h [hunk 1]
```diff
@@ -20,7 +20,7 @@
 # include <stdbool.h>
 # include <limits.h>
 
-# ifdef WITH_OPEN
```
Comment (Alexdelia) [inline libft/inc/libft.h]: libft doesn't pass the `norminette` version `3.3.55`

File: libft/inc/libft.h [hunk 2]
```diff
@@ -175,7 +183,7 @@ t_vec	*
 
 char		*get_next_line(int fd);
 
-# ifdef WITH_OPEN
```
Comment (Alexdelia) [inline libft/inc/libft.h]: https://github.com/maaike-slt/42-minishell/pull/3/files#r1874941005

File: libft/src/2d/ft_2d_push.c [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+#include "libft.h"
+
+void	**ft_2d_push(void ***ptr, void *val)
+{
+	size_t	size;
+	void	**new;
+
+	size = ft_2d_size((const void **)*ptr);
+	new = (void **)malloc(sizeof(void *) * (size + 1));
```
Comment (Alexdelia) [inline libft/src/2d/ft_2d_push.c]: probably need `size + 2`

File: libft/src/mem/ft_2d_drop.c
Comment (Alexdelia) [inline libft/src/mem/ft_2d_drop.c:1]: not supposed to have both `/mem/` and `/2d/`

File: src/builtin/b_exec.c
Comment (Alexdelia) [inline src/builtin/b_exec.c:1]: you can keep calling it `exec.c` instead of `b_exec`

File: src/builtin/cd.c [hunk 4]
```diff
@@ -6,25 +6,24 @@
...
+int	cd(t_args *arg)
 {
 	int	ret;
 
 	ret = 0;
-	if (argc > 2)
-		return (error(argv[0], "too many arguments"), EX_ERR);
-	if (argc == 1)
+	if (arg->argc > 2)
+		return (error(arg->argv[0], "too many arguments"), D_ERROR);
```
Comment (Alexdelia) [inline src/builtin/cd.c]:
```suggestion
		return (error(arg->argv[0], "too many arguments"), EX_ERR);
```

`D_XXX` is only for the `dispatch`

all `builtin` are like binary, they return an `exit code` that is `EX_XXX`

File: src/builtin/env.c [hunk 5]
```diff
@@ -17,12 +17,10 @@ int	env(char **envp)
...
 	i = 0;
-	while (envp[i] && envp[i + 1] != NULL)
+	while (envp[i])
 	{
 		ft_putendl_fd(envp[i], STDOUT_FILENO);
 		i++;
 	}
-	if (envp[i])
-		ft_putendl_fd(envp[i], STDOUT_FILENO);
 	return (EX_OK);
```
Comment (Alexdelia) [inline src/builtin/env.c:25]: great!

Comment (maaike-slt) [reply src/builtin/env.c:25]: <3

File: libft/src/mem/ft_2d_drop.c
Comment (maaike-slt) [reply libft/src/mem/ft_2d_drop.c:1]: :+1:

File: libft/src/2d/ft_2d_push.c (hunk 3 above)
Comment (maaike-slt) [reply libft/src/2d/ft_2d_push.c]: :+1:

File: libft/inc/libft.h (hunk 1 above)
Comment (maaike-slt) [reply libft/inc/libft.h]: :+1:

**Review by Alexdelia — APPROVED** [review]

**Pull request merged**